crates/miden-objects/src/account/component/metadata/mod.rs

```diff
@@ -0,0 +1,249 @@
+    #[cfg_attr(feature = "std", serde(rename = "storage"))]
+    storage_schema: AccountStorageSchema,
+}
+
+impl AccountComponentMetadata {
+    /// Create a new [AccountComponentMetadata].
+    ///
+    /// # Errors
+    ///
+    /// - If the schema contains invalid slot definitions.
+    /// - If the schema contains duplicate slot names.
+    /// - If the schema contains duplicate init value names.
+    pub fn new(
+        name: String,
+        description: String,
+        version: Version,
+        targets: BTreeSet<AccountType>,
+        storage_schema: AccountStorageSchema,
+    ) -> Result<Self, AccountComponentTemplateError> {
+        let component = Self {
+            name,
+            description,
+            version,
+            supported_types: targets,
+            storage_schema,
+        };
+        component.validate()?;
+        Ok(component)
+    }
+
+    /// Returns the init-time values required to instantiate this schema.
+    ///
+    /// These values are used for initializing storage slot values or storage map entries. For a
+    /// full example, refer to the docs for [AccountComponentMetadata].
+    ///
+    /// Types for returned init values are inferred based on their location in the storage layout.
+    pub fn init_value_requirements(&self) -> BTreeMap<StorageValueName, InitValueRequirement> {
+        let mut requirements = BTreeMap::new();
+        for (name, requirement) in self.storage_schema.init_value_requirements() {
+            requirements.insert(name, requirement);
+        }
+
+        requirements
+    }
+
+    /// Returns the name of the account component.
+    pub fn name(&self) -> &str {
+        &self.name
+    }
+
+    /// Returns the description of the account component.
+    pub fn description(&self) -> &str {
+        &self.description
+    }
+
+    /// Returns the semantic version of the account component.
+    pub fn version(&self) -> &Version {
+        &self.version
+    }
+
+    /// Returns the account types supported by the component.
+    pub fn supported_types(&self) -> &BTreeSet<AccountType> {
+        &self.supported_types
+    }
+
+    /// Returns the storage schema of the component.
+    pub fn storage_schema(&self) -> &AccountStorageSchema {
+        &self.storage_schema
+    }
+
+    /// Validate the [AccountComponentMetadata].
+    ///
+    /// # Errors
+    ///
+    /// - If the schema contains invalid slot definitions.
+    /// - If the schema contains duplicate init value names.
+    fn validate(&self) -> Result<(), AccountComponentTemplateError> {
+        self.storage_schema.validate()?;
+
+        // Check for duplicate init value names.
+        let mut seen_init_value_names = BTreeSet::new();
+        for (name, _) in self.storage_schema.init_value_requirements() {
+            if !seen_init_value_names.insert(name.clone()) {
+                return Err(AccountComponentTemplateError::DuplicateInitValueName(name));
+            }
+        }
```

Comment:
This shouldn't be possible, right? We are guaranteed that all names are unique because storage schema is based on a `BTreeMap` - correct?